### DISCUSSION TITLE
Request: more modifiers

### DISCUSSION THREAD

**Discussion by kontofikcyjne** [nikitabobko/AeroSpace#997] [feature-ideas]
Would it be possible to add more modifier keys? For example caps lock, § or fn?

**Comment by kontofikcyjne** [comment]
I think this should be added in this file /AeroSpace/Sources/AppBundle/config/keysMap.swift

**Comment by noahmayr** [comment]
Allowing to individually target left/right shift/alt/cmd would also be helpful imo. There are quite special chars I need alt for and I don't enjoy having to do cramp inducing keyboard acrobatics to do window management.

**Comment by blindfox0** [comment]
> Allowing to individually target left/right shift/alt/cmd would also be helpful imo. There are quite special chars I need alt for and I don't enjoy having to do cramp inducing keyboard acrobatics to do window management.

Exactly! I'm from poland so i have to use alt for letters like ą ć ś etc. Possibility of set for example only left alt for mod key will be great.

**Comment by anschuetz** [comment]
> Allowing to individually target left/right shift/alt/cmd would also be helpful imo. There are quite special chars I need alt for and I don't enjoy having to do cramp inducing keyboard acrobatics to do window management.

This would be #VERY# important: In Germany eg. "@" is alt-L so the window-navigation fails. I could do right-alt-L to @ and left-alt-L to navigate.

**Comment by bcremer** [comment]
A common workaround for non US keyboards is to use a software like https://hyperkey.app/ to map `alt-ctrl-cmd` to the capslock key.

I'm using `alt-ctrl-cmd` and `alt-ctrl-cmd-shift` as my primary modifier-prefix for all shortcuts in aerospace.

**Comment by anschuetz** [comment]
Ok, this brings me to another idea:
I'll try to use skhd (which supports lalt for the left alt key) and launch the aerospace-functions via cli. 
[5mins later]
cool. This works for me.

.skhdrc:
lalt - h : aerospace focus left
...

Thanks a lot!
Jesko

**Comment by jo-hb** [comment]
> > Allowing to individually target left/right shift/alt/cmd would also be helpful imo. There are quite special chars I need alt for and I don't enjoy having to do cramp inducing keyboard acrobatics to do window management.
> 
> This would be #VERY# important: In Germany eg. "@" is alt-L so the window-navigation fails. I could do right-alt-L to @ and left-alt-L to navigate.

It isn't only the alt-L key. It's also alt-5 = [ , alt-8 = { etc. that are blocked due to the absence an alternative availability of "fn" which would work perfectly on a German keyboard layout. Trying to workaround with skhd is also a bit tricky when it comes to the use of different modes. Adding the "fn" key would at least solve most of the problems for me.

**Reply by hajiboy95** [reply]
I'm using a German keyboard layout, but the key-mapping preset is set to `qwerty`. Is there a clean way to switch this to `qwertz` in QMK/ZMK? Otherwise, characters like `/`, `,`, `;`, etc., are all mismatched compared to the physical key positions.

Is the recommended way to handle this via `[key-mapping.key-notation-to-key-code]`, or is there a better method?

Also, I’ve ordered a split 40-key Corne keyboard and will be flashing it with QMK/ZMK. How does key mapping typically work in that case? Do I need to manually define every key position, or does the firmware handle layout abstraction in some way? Since there's no direct mapping from a full QWERTY/QWERTZ to 40 keys.

Thanks in advance

**Comment by wesbos** [comment]
Would like a bit more flexibility as well:

Would be nice to specify between left and right ⌘, ^, ⇧ and ⌥

Would be nice to use home, page up, page down, end, and delete as well - I often use these keys as a second ← ↑ ↓ → keypad

**Comment by babytof** [comment]
we have the same exact need for french keyboard where alt is needed for special chars like [ { } ] or | (alt+shift+L)
I already use karabiner to map Caps Lock to Esc and Ctrl, so I cannot use @bcremer proposal :/

**Comment by phischdev** [comment]
Is there a way to alias a combination of modifiers?

Such that I could write something like
`mod1 ~ cmd-option-command`
`mod1 = workspace 1`
`mod1-shift = move-to-workspace 1`
?

**Comment by hxhelm** [comment]
This is definitely a blocker for me, would love to try this out but having no support for differentiating right option from left option does not work with my setup.

I have tried implementing this and got the detection between right & left for the modifiers working, but got stuck because I ended up having to listen to both variants anyway.

The current limitation seems to be that [HotKey](https://github.com/soffes/HotKey) does not support that distinction.

**Comment by szymon3** [comment]
I managed to get it to work for Polish characters by removing workspace bindings that interfere with my special characters (e.g. alt+a, alt+c).

EDIT: Actually, not really. I forgot about ł character, which is alt (right option) + L

**Comment by thomdabrowski** [comment]
Blocker for me as well :/

**Comment by techeddie** [comment]
+1 add keys: 
home, page up, page down, end, and delete

**Comment by buzzedowl** [comment]
yes caps lock as a modifier would be great

**Comment by enzoenrico** [comment]
@nikitabobko please... :(

**Comment by ezadEzanee** [comment]
I had the same limitation, but no more. I am using Karabiner to control AeroSpace 99%. My shortcut key defined in aerospace.toml has only 3 lines

cmd-h = []                                                    # Disable "hide application"
cmd-alt-h = []                                                # Disable "hide others"
alt-shift-esc = ['reload-config', 'mode main']

the rest are all in Karabiner. It took me quite sometime to set it up in Karabiner, but the grind is definitely worth it. You can setup layering and assign whatever key you want as lead/shortcut prefix key.

for example i setup my Tab as the leadkey for anything Windows Management. 

Tab - A is for Application, from there you can define additional layers for example B for browsers and under B theres Edge, Safari, Firefox, Zen etc etc.
Tab - W : Workspace manaegment 
Tab - M : Move layer
Tab - G : Join layer
Tab - W - R for resize layer. 

Depending on what layer i am in H, J ,K, L will either move, join or focus window left, down, up or right. 

The best part, Tab still remains available for me to indent to the right or left. I have 0 shortcut clashes eversince going this route.

**Reply by ezadEzanee** [reply]
the layer concept is similar to binding mode implemented in AeroSpace. The difference is how/when you exit the layer/binding. 

In AeroSpace one would define it by chaining ", 'mode main']" after the command or by executing another shortcut that specifically brings you back to the main bindings. So let's say i want to open 3 different applications. using AeroSpace keybinding would require 9 keys to be punched in. e.g. Opt-A-B, Opt-A-M and Opt-A-S. can also configure it to remain in Application binding mode but I have to remember to exit the Application binding mode. or else i would be launching alot of applications once i start typing. 

In Karabiner, the same action only requires 5 keys. Tab(hold) - A - B,M,S and all 3 apps are opened. And I simply need to lift my finger off the Tab key. Less keystrokes for the same actions

**Reply by anschuetz** [reply]
Hi @ezadEzanee, 
your approach sounds very promising to me. I'd like to try it. 
How did you configure this "layering"?
Could you export your config and provide it to me in some way? That would be very nice!

Thank you in advance!
best regards Jesko

**Comment by Cuhadari-Deniz** [comment]
I would love to have something like F13 as my main modifier key for my window manager. That way i would never run into overlapping keymaps with other applications and i would simply map F13 via karabiner or kanata.

**Comment by Fanzzzd** [comment]
I recommend using Hammerspoon to implement custom configurations.
You may refer to my existing code, though please note its complexity: I have utilized Karabiner to remap the Eisu key on a Japanese keyboard to a function key.

```lua
local obj = {}
obj.name = "AeroSpace"
obj.version = "1.0"
obj.license = "MIT"

obj.aerospace_path = "/opt/homebrew/bin/aerospace"

function obj:aerospace(command)
  hs.execute(self.aerospace_path .. " " .. command)
end

function obj:init()
    local RCMD_MASK = 16
    local LCMD_MASK = 8
    local SHIFT_RAW_MASK = 2 | 4
    local ALT_RAW_MASK = 32 | 64
    local CTRL_RAW_MASK = 1 | 8192

    local rcmd_key_map = {
        h = function() self:aerospace('focus --boundaries-action wrap-around-the-workspace left') end,
        j = function() self:aerospace('focus --boundaries-action wrap-around-the-workspace down') end,
        k = function() self:aerospace('focus --boundaries-action wrap-around-the-workspace up') end,
        l = function() self:aerospace('focus --boundaries-action wrap-around-the-workspace right') end,
        ['1'] = function() self:aerospace('workspace 1') end,
        ['2'] = function() self:aerospace('workspace 2') end,
        ['3'] = function() self:aerospace('workspace 3') end,
        ['4'] = function() self:aerospace('workspace 4') end,
        ['5'] = function() self:aerospace('workspace 5') end,
        ['6'] = function() self:aerospace('workspace 6') end,
        ['7'] = function() self:aerospace('workspace 7') end,
        ['8'] = function() self:aerospace('workspace 8') end,
        ['9'] = function() self:aerospace('workspace et1') end,
        ['0'] = function() self:aerospace('workspace et2') end,
        ['/'] = function() self:aerospace('layout horizontal vertical') end,
        e = function() self:aerospace('balance-sizes') end,
        g = function() hs.execute("open -a Ghostty") end,
        f = function() self:aerospace('fullscreen') end
    }

    local rcmd_shift_key_map = {
        h = function() self:aerospace('move left') end,
        j = function() self:aerospace('move down') end,
        k = function() self:aerospace('move up') end,
        l = function() self:aerospace('move right') end,
        left = function() self:aerospace('move left') end,
        down = function() self:aerospace('move down') end,
        up = function() self:aerospace('move up') end,
        right = function() self:aerospace('move right') end,
        p = function() self:aerospace('move-node-to-workspace prev') end,
        n = function() self:aerospace('move-node-to-workspace next') end,
        ['1'] = function() self:aerospace('move-node-to-workspace --focus-follows-window 1') end,
        ['2'] = function() self:aerospace('move-node-to-workspace --focus-follows-window 2') end,
        ['3'] = function() self:aerospace('move-node-to-workspace --focus-follows-window 3') end,
        ['4'] = function() self:aerospace('move-node-to-workspace --focus-follows-window 4') end,
        ['5'] = function() self:aerospace('move-node-to-workspace --focus-follows-window 5') end,
        ['6'] = function() self:aerospace('move-node-to-workspace --focus-follows-window 6') end,
        ['7'] = function() self:aerospace('move-node-to-workspace --focus-follows-window 7') end,
        ['8'] = function() self:aerospace('move-node-to-workspace --focus-follows-window 8') end,
        ['9'] = function() self:aerospace('move-node-to-workspace --focus-follows-window et1') end,
        ['0'] = function() self:aerospace('move-node-to-workspace --focus-follows-window et2') end,
        c = function() self:aerospace('reload-config') end,
        s = function() self:aerospace('enable off') end,
        f = function() self:aerospace('layout floating tiling') end
    }

    local rcmd_ctrl_key_map = {
        h = function() self:aerospace('join-with left') end,
        j = function() self:aerospace('join-with down') end,
        k = function() self:aerospace('join-with up') end,
        l = function() self:aerospace('join-with right') end,
        f = function() self:aerospace('flatten-workspace-tree') end
    }

    self.eventtap = hs.eventtap.new({hs.eventtap.event.types.keyDown, hs.eventtap.event.types.keyUp}, function(e)
        local raw_flags = e:rawFlags()

        if not ((raw_flags & RCMD_MASK > 0) and (raw_flags & LCMD_MASK == 0)) then
            return false
        end

        local key_char = hs.keycodes.map[e:getKeyCode()]
        local event_type = e:getType()
        local has_shift = (raw_flags & SHIFT_RAW_MASK) > 0
        local has_alt = (raw_flags & ALT_RAW_MASK) > 0
        local has_ctrl = (raw_flags & CTRL_RAW_MASK) > 0
        local no_other_mods = not has_alt and not has_ctrl

        if event_type ~= hs.eventtap.event.types.keyDown then
            return false
        end

        if has_shift and no_other_mods then
            local fn = rcmd_shift_key_map[key_char]
            if fn then
                fn()
                return true
            end
        elseif has_ctrl and not has_shift and not has_alt then
            local fn = rcmd_ctrl_key_map[key_char]
            if fn then
                fn()
                return true
            end
        elseif not has_shift and no_other_mods then
            local fn = rcmd_key_map[key_char]
            if fn then
                fn()
                return true
            end
        end

        return false
    end)
    self.eventtap:start()
end

function obj:stop()
    if self.eventtap then
        self.eventtap:stop()
        self.eventtap = nil
    end
end

return obj
```

Despite this complexity, the underlying logic remains applicable.
In essence, you can bind any specific keys through Hammerspoon and pass arguments to execute window management operations.
